File: ConsoleApplication1/ClassLibrary2/DerivedClass2.cpp
#include "stdafx.h"
#include "DerivedClass2.h"

using namespace System;

void ClassLibrary2::DerivedClass2::Function()
{
	Console::WriteLine(L"[IN] DerivedClass2::Function is called.");
	ClassLibrary1::BaseBaseClass::Function();
	Console::WriteLine(L"[OUT] DerivedClass2::Function is called.");
}

File: ConsoleApplication1/ClassLibrary3/DerivedClass2.cs
using System;
using System.Reflection.Emit;
using ClassLibrary1;

namespace ClassLibrary3
{
    public class DerivedClass2 : BaseClass
    {
        private static readonly Action<BaseBaseClass> _callFunction;

        static DerivedClass2()
        {
            var dm = new DynamicMethod("", typeof(void), new[] { typeof(BaseBaseClass) }, typeof(DerivedClass2));
            var il = dm.GetILGenerator();

            il.Emit(OpCodes.Ldarg_0);
            il.Emit(OpCodes.Call, typeof(BaseBaseClass).GetMethod("Function"));
            il.Emit(OpCodes.Ret);

            _callFunction = (Action<BaseBaseClass>)dm.CreateDelegate(typeof(Action<BaseBaseClass>));
        }

        public override void Function()
        {
            Console.WriteLine($"[IN] {nameof(DerivedClass2)}.{nameof(Function)} is called.");
            _callFunction?.Invoke(this);
            Console.WriteLine($"[OUT] {nameof(DerivedClass2)}.{nameof(Function)} is called.");
        }
    }
}
